ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/PeerDiscoveryAgent.java

```diff
@@ -182,23 +182,28 @@ protected boolean validatePacketSize(final int packetSize) {
   }
 
   protected void handleIncomingPacket(final Endpoint sourceEndpoint, final Packet packet) {
-    OptionalInt tcpPort = OptionalInt.empty();
-    if (packet.getPacketData(PingPacketData.class).isPresent()) {
-      final PingPacketData ping = packet.getPacketData(PingPacketData.class).orElseGet(null);
-      if (ping != null && ping.getFrom() != null && ping.getFrom().getTcpPort().isPresent()) {
-        tcpPort = ping.getFrom().getTcpPort();
-      }
-    }
+    final int port = sourceEndpoint.getUdpPort();
+    final int tcpPort =
+        packet
+            .getPacketData(PingPacketData.class)
+            .flatMap(PingPacketData::getFrom)
+            .flatMap(
+                fromEndpoint -> {
+                  final OptionalInt maybePort = fromEndpoint.getTcpPort();
+                  return maybePort.isPresent()
```

Comment:
perhaps map instead of isPresent ? otherOptional : empty

Actually `OptionalInt` doesn't have `map`.  That's why I hate it and use `Optional<Integer>`

Reply:
Totally agree, I already have a separate PR that I'll put up after this that includes the change.